### PR TITLE
Make float16 training work for gemma4, qwen3_moe and glm4_moe

### PR DESCRIPTION
**Pull request by danielhanchen**
## Summary

Makes `dtype=torch.float16` training work for three MoE architectures that silently NaN today: `gemma4` (Gemma-4), `qwen3_moe` (Qwen3-30B-A3B) and `glm4_moe` (GLM-4.x MoE). They now behave the same as `gpt_oss` / `qwen3_5` / `gemma3`.

Loaded with `dtype=torch.float16` these models keep a finite loss but NaN the grad_norm (mostly on the first step, gemma4 more often). It is a backward gradient overflow along the residual stream: forward activations peak around 350, far below fp16's 65504, and the same runs are fully finite in bf16. The nan-grad step gets skipped by the optimizer, so training limps along but is unreliable.

## Changes

- `model_lists.py`: add `gemma4`, `glm4_moe` and `qwen3_moe` to `FORCE_FLOAT32`, so a float16 request loads bf16 weights and sets `do_forced_float32` (as it already does for gpt_oss / qwen3_5 / gemma3).
- `glm4_moe` needs nothing else: its modules are dtype-consistent, so the bf16 load alone trains finite.
- `gemma4` and `qwen3_moe` use QK-norm, so `do_forced_float32`'s generic float32 upcast puts q/k in float32 while v stays fp16 and SDPA raises `expected ... the same dtype`. This adds targeted per-module float32 patches (`gemma4_float32.py`, `qwen3_moe_float32.py`), gated on `UNSLOTH_FORCE_FLOAT32`, that keep only the precision-sensitive ops in float32:
  - the residual stream (gemma4 via its scaled word embedding; qwen3 via the decoder layer, since its embedding is unscaled),
  - the RMSNorm internals (returning fp16, clamped to the fp16 range),
  - the attention core (q/k/v and scores all float32, consistent dtype, output down-cast to fp16 for `o_proj`).

  Every projection, MoE expert and router stays fp16/4bit, so this keeps the fast path fast (no whole-model float32). The patches are direct analogs of the existing `gemma3` patches in `gemma.py` and no-op entirely when `UNSLOTH_FORCE_FLOAT32` is 0.

## Validation (single B200, 4bit QLoRA, transformers 5.5)

- `gemma4`: float16 request trains FINITE over 15 steps (no crash, no NaN); bf16 unchanged and the losses track float16 almost exactly (step 1 0.899 vs 0.901; step 15 0.628 vs 0.619).
- `qwen3_moe`: float16 FINITE over 15 steps (grad_norm 0.8-1.9); bf16 unchanged.
- `glm4_moe`: float16 FINITE over 15 steps (grad_norm 0.17-0.38).
- Targeted upcast confirmed: heavy matmuls and experts stay fp16/uint8 (zero float32 parameters); only the residual/norm/attention-core activations are float32.
- The bf16 path is provably untouched (all patches gate on `UNSLOTH_FORCE_FLOAT32 == "1"`), so existing behavior, including the gemma4 sliding-window attention path, is unchanged.

## Notes

- `deepseek_v4` is another MoE that likely needs the same treatment, but transformers 5.5 does not register `deepseek_v4`, so it cannot be validated through the standard loader yet. Left as a follow-up.
- The mirrored fallback list in `unsloth/models/loader.py` (used only if the unsloth_zoo import fails) is updated in a companion change on the unsloth side.

### CODE REVIEW

**Comment by danielhanchen** [comment]
@codex review

**Comment by danielhanchen** [comment]
The remaining suggestions are the `pass` separators used throughout the temporary_patches modules, so I kept them for consistency with the surrounding code. No functional issues were raised.

File: unsloth_zoo/temporary_patches/gemma4_float32.py [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+            scale = self.scaling,
+            enable_gqa = getattr(self, "num_key_value_groups", 1) != 1,
+        )
+
+        # (b, h, q, d) -> (b, q, h*d), down-cast to fp16 for o_proj.
+        attn_output = attn_output.transpose(1, 2).contiguous().reshape(*input_shape, -1)
+        attn_output = attn_output.to(torch.float16)
+        attn_output = self.o_proj(attn_output)
+        return attn_output, None
+    pass
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/temporary_patches/gemma4_float32.py:239]: These standalone pass separators are a deliberate convention throughout temporary_patches (gpt_oss.py alone has dozens), so I am keeping this one for consistency with the surrounding modules. No functional change.

**Comment by danielhanchen** [comment]
@codex review

File: unsloth_zoo/temporary_patches/gemma4_float32.py [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+    scaled_dot_product_attention = torch.nn.functional.scaled_dot_product_attention
+    scaled_dot_product_attention = torch.compiler.disable(scaled_dot_product_attention, recursive = True)
+
+    def forward(
+        self,
+        hidden_states,
+        position_embeddings = None,
+        attention_mask = None,
+        past_key_values = None,
+        **kwargs,
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/temporary_patches/gemma4_float32.py:114]: Good catch. The forward now accepts shared_kv_states and reads/writes it keyed by layer_type, matching the native 5.5.2+ protocol, with the 5.5.0-5.5.1 carrier path kept as fallback. Verified producer/consumer round trip on a kv-shared pair.

File: unsloth_zoo/temporary_patches/qwen3_moe_float32.py [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+        # Core attention in float32 (consistent q/k/v dtype -> no SDPA mismatch).
+        attn_mask = attention_mask
+        if isinstance(attn_mask, torch.Tensor) and attn_mask.dtype != torch.bool:
+            attn_mask = attn_mask.to(torch.float32)
+        is_causal = attn_mask is None and query_states.shape[2] > 1 and getattr(self, "is_causal", True)
+        attn_output = scaled_dot_product_attention(
+            query_states,
+            key_states,
+            value_states,
+            attn_mask = attn_mask,
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/temporary_patches/qwen3_moe_float32.py]: Fixed: both float32 attention replacements now dispatch flex_attention through the attention interface like the Gemma3 patch, raw SDPA is only used for the other implementations.

File: unsloth_zoo/temporary_patches/gemma4_float32.py [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+        # Core attention in float32. Gemma4 uses scaling = 1.0 (QK-norm bakes the scale in).
+        attn_mask = attention_mask
+        if isinstance(attn_mask, torch.Tensor) and attn_mask.dtype != torch.bool:
+            attn_mask = attn_mask.to(torch.float32)
+        is_causal = attn_mask is None and query_states.shape[2] > 1 and getattr(self, "is_causal", True)
+        attn_output = scaled_dot_product_attention(
+            query_states,
+            key_states,
+            value_states,
+            attn_mask = attn_mask,
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/temporary_patches/gemma4_float32.py]: Fixed together with the Qwen3-MoE one: flex masks now go through the attention interface.

File: unsloth_zoo/temporary_patches/qwen3_moe_float32.py [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+            past_key_values = past_key_values,
+            use_cache = use_cache,
+            position_embeddings = position_embeddings,
+            **kwargs,
+        )  # -> fp16
+        hidden_states = residual + hidden_states  # fp32 + fp16 -> fp32
+
+        residual = hidden_states  # fp32
+        hidden_states = self.post_attention_layernorm(hidden_states)  # -> fp16
+        hidden_states = self.mlp(hidden_states)  # MoE / MLP -> fp16
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/temporary_patches/qwen3_moe_float32.py:87]: Fixed: the decoder replacement now unpacks tuple MLP outputs the same way the upstream decoder does, so the old-transformers sparse block contract works under UNSLOTH_FORCE_FLOAT32.

**Comment by danielhanchen** [comment]
@codex review

File: unsloth_zoo/temporary_patches/qwen3_moe_float32.py [hunk 6]
```diff
@@ -0,0 +1,209 @@
...
+        # q/k: fp16 proj -> fp16 q_norm/k_norm (returns fp16) -> float32 -> (b, h, s, d) -> RoPE (float32)
+        query_states = self.q_norm(self.q_proj(hidden_states).view(hidden_shape)).to(torch.float32).transpose(1, 2)
+        key_states   = self.k_norm(self.k_proj(hidden_states).view(hidden_shape)).to(torch.float32).transpose(1, 2)
+        value_states = self.v_proj(hidden_states).view(hidden_shape).to(torch.float32).transpose(1, 2)
+
+        # Qwen3 applies RoPE to both q and k together, unsqueeze_dim = 1 (post-transpose layout).
+        query_states, key_states = apply_rotary_pos_emb(query_states, key_states, cos, sin)
+
+        if past_key_values is not None:
+            key_states, value_states = past_key_values.update(key_states, value_states, self.layer_idx)
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/temporary_patches/qwen3_moe_float32.py]: Fixed: the update call now forwards cache_kwargs with sin, cos and cache_position, which 4.57.x static caches need; 5.x layer caches absorb the extra dict.

File: unsloth_zoo/temporary_patches/qwen3_moe_float32.py [hunk 7]
```diff
@@ -0,0 +1,209 @@
...
+        attn_mask = attention_mask
+        if isinstance(attn_mask, torch.Tensor) and attn_mask.dtype != torch.bool:
+            attn_mask = attn_mask.to(torch.float32)
+        attn_impl = getattr(self.config, "_attn_implementation", "sdpa")
+        if _UNSLOTH_FLEX_ATTENTION_DISABLED:
+            attn_impl = "sdpa"
+        if attn_impl == "flex_attention":
+            # Flex builds a BlockMask that raw SDPA cannot consume; dispatch to the
+            # interface (returns (b, q, h, d), already transposed).
+            attention_interface = ALL_ATTENTION_FUNCTIONS[attn_impl]
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/temporary_patches/qwen3_moe_float32.py]: Fixed: non-SDPA backends now dispatch through ALL_ATTENTION_FUNCTIONS with fp16 inputs (flash accumulates in float32 internally), so the 2-D padding mask keeps its native semantics.

File: unsloth_zoo/temporary_patches/gemma4_float32.py [hunk 8]
```diff
@@ -0,0 +1,217 @@
...
+        attn_mask = attention_mask
+        if isinstance(attn_mask, torch.Tensor) and attn_mask.dtype != torch.bool:
+            attn_mask = attn_mask.to(torch.float32)
+        attn_impl = getattr(self.config, "_attn_implementation", "sdpa")
+        if _UNSLOTH_FLEX_ATTENTION_DISABLED:
+            attn_impl = "sdpa"
+        if attn_impl == "flex_attention":
+            # Flex builds a BlockMask that raw SDPA cannot consume; dispatch to the
+            # interface (returns (b, q, h, d), already transposed).
+            attention_interface = ALL_ATTENTION_FUNCTIONS[attn_impl]
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/temporary_patches/gemma4_float32.py]: Fixed the same way for Gemma4: registered backends get dispatched through the interface with fp16 inputs; raw SDPA is now only used for sdpa itself or unknown backends.

File: unsloth_zoo/temporary_patches/qwen3_moe_float32.py [hunk 9]
```diff
@@ -0,0 +1,209 @@
...
+                scale = self.scaling,
+                enable_gqa = getattr(self, "num_key_value_groups", 1) != 1,
+            )
+            attn_output = attn_output.transpose(1, 2)  # (b, h, q, d) -> (b, q, h, d)
+
+        # (b, q, h, d) -> (b, q, h*d), down-cast to fp16 for o_proj.
+        attn_output = attn_output.contiguous().reshape(*input_shape, -1)
+        attn_output = attn_output.to(torch.float16)
+        attn_output = self.o_proj(attn_output)
+        return attn_output, None
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/temporary_patches/qwen3_moe_float32.py]: Added an eager branch that computes attention in float32 and returns the weights, restoring the module-level eager contract. Note the qwen3_moe decoder discards the weights on both 4.57.x and 5.x, so model-level outputs were never affected.

File: unsloth_zoo/temporary_patches/gemma4_float32.py [hunk 10]
```diff
@@ -0,0 +1,217 @@
...
+                scale = self.scaling,
+                enable_gqa = getattr(self, "num_key_value_groups", 1) != 1,
+            )
+            attn_output = attn_output.transpose(1, 2)  # (b, h, q, d) -> (b, q, h, d)
+
+        # (b, q, h, d) -> (b, q, h*d), down-cast to fp16 for o_proj.
+        attn_output = attn_output.contiguous().reshape(*input_shape, -1)
+        attn_output = attn_output.to(torch.float16)
+        attn_output = self.o_proj(attn_output)
+        return attn_output, None
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/temporary_patches/gemma4_float32.py]: Same eager branch added for Gemma4; weights are computed in float32 and returned. Gemma4 upstream has no output_attentions collection, so this only affects direct module calls.

File: unsloth_zoo/model_lists.py [hunk 11]
```diff
@@ -46,6 +46,9 @@
     "gemma3,",     # trailing comma is a substring-path delimiter for the CUDA loader
     "gemma3text",  # EmbeddingGemma / standalone text-only Gemma3 (config: "gemma3_text")
     "gemma3n",
+    "gemma4",      # substring-matches "gemma4" and "gemma4_text"; fp16 NaNs the grad_norm in the backward (targeted float32 via gemma4_float32.py)
+    "glm4_moe",    # GLM-4.x MoE (glm4_moe / glm4_moe_lite): fp16 NaNs grad_norm; bf16 load is dtype-consistent
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/model_lists.py]: Added gemma4text and glm4_moe_lite entries so the MLX exact-match helper covers those config model_types; the CUDA substring path already matched them via the base entries.

**Comment by danielhanchen** [comment]
@codex review

**Comment by danielhanchen** [comment]
@codex review

**Comment by danielhanchen** [comment]
### float16 before / after

Quick summary of what this PR fixes. With `dtype=torch.float16` these three MoE models used to keep a finite loss but NaN the grad_norm (mostly on the very first step, gemma4 most often), so the optimizer skipped the step and training limped along unreliably. After the fix they train finite and behave like the bf16 runs.

| model | fp16 grad_norm before | fp16 grad_norm after | note |
| --- | --- | --- | --- |
| gemma4 | NaN on step ~1 (the most frequent of the three) | finite over 15 steps, no NaN; loss tracks bf16 almost exactly (step 1 0.899 vs 0.901, step 15 0.628 vs 0.619) | uses QK-norm, needs the targeted float32 patch |
| qwen3_moe (Qwen3-30B-A3B) | NaN grad_norm, usually the first step | finite over 15 steps, grad_norm 0.8-1.9 | uses QK-norm, needs the targeted float32 patch |
| glm4_moe (GLM-4.x MoE) | NaN grad_norm, usually the first step | finite over 15 steps, grad_norm 0.17-0.38 | dtype-consistent, the bf16 load alone trains finite |

Root cause: a backward gradient overflow along the residual stream. Forward activations peak around 350, far below fp16's 65504, but the backward pass overflows, which is why the loss stays finite while the grad_norm goes NaN.

The same runs were always fully finite in bf16, which is what the fix leans on: a float16 request now loads bf16 weights and sets `do_forced_float32`, with targeted per-module float32 patches on the residual/norm/attention-core only for the two QK-norm models. Everything gates on `UNSLOTH_FORCE_FLOAT32 == "1"`, so the bf16 path is untouched.

Note on the numbers: the "after" grad_norm ranges and the gemma4 loss pairs are measured (single B200, 4bit QLoRA, transformers 5.5, 15 steps); the "before" column is qualitative, since the pre-fix failure mode was a NaN grad_norm rather than a stable value to tabulate.
